crates/telemetry/src/traces.rs

```diff
@@ -0,0 +1,180 @@
+
+    // This will configure the exporter based on the OTEL_EXPORTER_* environment variables. We
+    // currently default to using the HTTP exporter but in the future we could select off of the
+    // combination of OTEL_EXPORTER_OTLP_PROTOCOL and OTEL_EXPORTER_OTLP_TRACES_PROTOCOL to
+    // determine whether we should use http/protobuf or grpc.
+    let mut exporter = opentelemetry_otlp::new_exporter().http();
+
+    // This mitigation was taken from https://github.com/neondatabase/neon/blob/main/libs/tracing-utils/src/lib.rs
+    //
+    // opentelemetry-otlp v0.15.0 has a bug in how it uses the
+    // OTEL_EXPORTER_OTLP_ENDPOINT env variable. According to the
+    // OpenTelemetry spec at
+    // <https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp>,
+    // the full exporter URL is formed by appending "/v1/traces" to the value
+    // of OTEL_EXPORTER_OTLP_ENDPOINT. However, opentelemetry-otlp only does
+    // that with the grpc-tonic exporter. Other exporters, like the HTTP
+    // exporter, use the URL from OTEL_EXPORTER_OTLP_ENDPOINT as is, without
+    // appending "/v1/traces".
+    //
+    // See https://github.com/open-telemetry/opentelemetry-rust/pull/950
+    //
+    // Work around that by checking OTEL_EXPORTER_OTLP_ENDPOINT, and setting
+    // the endpoint url with the "/v1/traces" path ourselves. If the bug is
+    // fixed in a later version, we can remove this code. But if we don't
+    // remember to remove this, it won't do any harm either, as the crate will
+    // just ignore the OTEL_EXPORTER_OTLP_ENDPOINT setting when the endpoint
+    // is set directly with `with_endpoint`.
+    if std::env::var(OTEL_EXPORTER_OTLP_TRACES_ENDPOINT).is_err() {
+        if let Ok(mut endpoint) = std::env::var(OTEL_EXPORTER_OTLP_ENDPOINT) {
+            if !endpoint.ends_with('/') {
+                endpoint.push('/');
+            }
+            endpoint.push_str("v1/traces");
+            exporter = exporter.with_endpoint(endpoint);
+        }
+    }
+
+    let tracer = opentelemetry_otlp::new_pipeline()
+        .tracing()
+        .with_exporter(exporter)
+        .with_trace_config(
+            opentelemetry_sdk::trace::config().with_resource(Resource::new(service_metadata)),
+        )
+        .install_batch(opentelemetry_sdk::runtime::Tokio)?;
+
+    Ok(tracing_opentelemetry::layer()
+        .with_tracer(tracer)
+        .with_threads(false)
+        .with_filter(LevelFilter::from_level(Level::from_str(
+            config.otel_tracing_level.as_str(),
+        )?)))
+    // TODO: Set the filter level so only info spans from spin or spin children crates makes it
+    // TODO: Make sure it doesn't blow up to badly if the exporter doesn't exist
+}
+
+/// Injects the current W3C TraceContext into the provided request.
+pub fn inject_trace_context<'a>(req: impl Into<HeaderInjector<'a>>) {
```

Comment:
Might be nice to split everything from here down into a e.g. `propagation` submodule. :shrug: